pkg/storage/edv/restprovider.go

```diff
@@ -179,24 +175,20 @@ type restStore struct {
 }
 
 // v must be a marshalled EncryptedDocument.
+// Important: If updating an existing k, v pair, then the document ID in v will be replaced with the
+// existing encrypted document's ID.
 func (r *restStore) Put(k string, v []byte) error {
-	var encryptedDocument models.EncryptedDocument
-
-	err := json.Unmarshal(v, &encryptedDocument)
-	if err != nil {
-		return fmt.Errorf(failUnmarshalValueIntoEncryptedDocument, err)
+	// See if this key is already being used in an encrypted index tag in an existing document.
+	existingEDVDocumentID, err := r.retrieveEDVDocumentID(k)
+	if err != nil && !errors.Is(err, storage.ErrDataNotFound) {
+		return fmt.Errorf(failCheckForExistingEDVDocument, err)
 	}
 
-	indexedAttributeCollection, err := r.createIndexedAttributes(k)
+	// If existingEDVDocumentID was set, then this means that there is already an existing document that
+	// well get updated.
+	err = r.createEDVDocument(k, v, existingEDVDocumentID)
```

Comment:
trying to understand the logic here since `errors.Is(err, storage.ErrDataNotFound)` is not explicitly covered...

if `existingEDVDocumentID` is empty.. it's going to create a new edv document, otherwise it will fetch the existing document?

Reply:
@Baha-sk Yes, exactly.

Reply:
I think `r.createEDVDocument` is a bit confusing... it should be createOrUpdateEDVDocument ... or better do 2 functions, one create and one update... and add the `if existingEDVDocumentID == ""` here instead

Reply:
Actually, you're almost correct. If existingEDVDocumentID is empty, then it's going to update that document instead of creating a new one (the old document is never actually fetched though)

Reply:
I think `createOrUpdateEDVDocument` is a more accurate name.. but it's private.. so it should be fine

Reply:
Actually that's what I originally did - 2 separate functions. The problem was a lot of code duplication and it would have been multiple tests too... I struggled with the name a bit as well, as I was trying to keep it short as well based on previous feedback I got from @sandrask... What do you think about the name? Is there maybe some other nice, succinct way of describing this?

Reply:
createOrUpdateEDVDocument is short and clear

Reply:
OK - I'll fix this later in #2315. I'll be doing a big storage refactor there anyway. I linked to the comment here in that issue so I won't forget to do it.